File: backend/app/models/ticket.py
from uuid import UUID

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database.session import get_db

from app.repositories.ticket_repository import (
    TicketRepository,
)
from app.repositories.interaction_repository import (
    InteractionRepository,
)

from app.schemas.ticket import TicketResponse
from app.schemas.interaction import InteractionResponse
from app.schemas.ticket_from_interaction import (
    TicketFromInteractionCreate,
    TicketFromInteractionResponse,
)
from app.schemas.attach_interaction import (
    AttachInteractionRequest,
    AttachInteractionResponse,
)

from app.services.ticket_service import (
    TicketService,
)
from app.services.interaction_service import (
    InteractionService,
)
from app.services.inbox_ticket_service import (
    InboxTicketService,
)

router = APIRouter(
    prefix="/tickets",
    tags=["Tickets"],
)

# =========================================================
# Workflow 1
# Create Ticket From Inbox Interaction
# =========================================================

@router.post(
    "/from-interaction",
    response_model=TicketFromInteractionResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_ticket_from_interaction(
    request: TicketFromInteractionCreate,
    db: AsyncSession = Depends(get_db),
):

    ticket_repository = TicketRepository(db)

    interaction_repository = InteractionRepository(db)

    service = InboxTicketService(
        ticket_repository=ticket_repository,
        interaction_repository=interaction_repository,
    )

    return await service.create_ticket_from_interaction(
        request
    )


# =========================================================
# Workflow 2
# Attach Interaction To Existing Ticket
# =========================================================

@router.post(
    "/{ticket_id}/attach-interaction",
    response_model=AttachInteractionResponse,
    status_code=status.HTTP_200_OK,
)
async def attach_interaction_to_ticket(
    ticket_id: UUID,
    request: AttachInteractionRequest,
    db: AsyncSession = Depends(get_db),
):

    ticket_repository = TicketRepository(db)

    interaction_repository = InteractionRepository(db)

    service = InboxTicketService(
        ticket_repository=ticket_repository,
        interaction_repository=interaction_repository,
    )

    return await service.attach_to_existing_ticket(
        ticket_id=ticket_id,
        request=request,
    )


# =========================================================
# Ticket Timeline
# =========================================================

@router.get(
    "/{ticket_id}/interactions",
    response_model=list[InteractionResponse],
    status_code=status.HTTP_200_OK,
)
async def get_ticket_interactions(
    ticket_id: UUID,
    db: AsyncSession = Depends(get_db),
):

    interaction_repository = InteractionRepository(db)

    service = InteractionService(
        interaction_repository=interaction_repository,
    )

    return await service.get_ticket_interactions(
        ticket_id
    )


# =========================================================
# Get Ticket Details
# =========================================================

@router.get(
    "/{ticket_id}",
    response_model=TicketResponse,
    status_code=status.HTTP_200_OK,
)
async def get_ticket(
    ticket_id: UUID,
    db: AsyncSession = Depends(get_db),
):

    ticket_repository = TicketRepository(db)

    service = TicketService(
        ticket_repository=ticket_repository,
    )

    return await service.get_by_id(
        ticket_id
    )

File: frontend/src/api/ticket.ts
import { apiClient } from "./client";
import type {
  AttachInteractionRequest,
  AttachInteractionResponse,
  TicketActionResponse,
  TicketFromInteractionRequest,
  TicketFromInteractionResponse,
  TicketResponse,
  TicketUpdateRequest,
  TransferAgentRequest,
} from "@/types";

// GET /tickets?agent_name=...
export async function listTickets(agentName?: string): Promise<TicketResponse[]> {
  const { data } = await apiClient.get<TicketResponse[]>("/tickets", {
    params: agentName ? { agent_name: agentName } : undefined,
  });
  return data;
}

// POST /tickets/from-interaction
export async function createTicketFromInteraction(
  payload: TicketFromInteractionRequest
): Promise<TicketFromInteractionResponse> {
  const { data } = await apiClient.post<TicketFromInteractionResponse>(
    "/tickets/from-interaction",
    payload
  );
  return data;
}

// POST /tickets/{ticket_id}/attach-interaction
export async function attachInteractionToTicket(
  ticketId: string,
  payload: AttachInteractionRequest
): Promise<AttachInteractionResponse> {
  const { data } = await apiClient.post<AttachInteractionResponse>(
    `/tickets/${ticketId}/attach-interaction`,
    payload
  );
  return data;
}

// GET /tickets/{ticket_id}?agent_name=...
export async function getTicket(
  ticketId: string,
  agentName?: string
): Promise<TicketResponse> {
  const { data } = await apiClient.get<TicketResponse>(`/tickets/${ticketId}`, {
    params: agentName ? { agent_name: agentName } : undefined,
  });
  return data;
}

// PATCH /tickets/{ticket_id}
export async function updateTicket(
  ticketId: string,
  payload: TicketUpdateRequest,
  agentName?: string
): Promise<TicketResponse> {
  const { data } = await apiClient.patch<TicketResponse>(
    `/tickets/${ticketId}`,
    payload,
    { params: agentName ? { agent_name: agentName } : undefined }
  );
  return data;
}

// POST /tickets/{ticket_id}/transfer
export async function transferTicketAgent(
  ticketId: string,
  payload: TransferAgentRequest,
  agentName?: string
): Promise<TicketActionResponse> {
  const { data } = await apiClient.post<TicketActionResponse>(
    `/tickets/${ticketId}/transfer`,
    payload,
    { params: agentName ? { agent_name: agentName } : undefined }
  );
  return data;
}
